---
layout: default
title: About Me
---

I am half biologist, half programmer and half statistician, that makes me 1.5 scientist.. wait a minute.... 

Formally I am a researcher at a university, but informally I am just someone who enjoys learning new things and solving problems. Being a researcher, I am given the opportunity to do both, where I have to study and/or solve a problem, and most of the time in order to do so, I am required to learn new things, either technical methods or the background knowledge necessary. I aspire to be someone that can solve (almost) any problem using computer and mathematics. 

Below is my resume summary.

## Research Experience
#### Statistical Genetics, 2023 - Present
Multi-ancestry genetics research with the aim developing method that would harness the information from all samples including their ancestral genetic structure. This would be a method that represent all ancestries including admixed individuals (individual with mixed ancestry genetics, e.g. someone with a parent of east asian descent and a parent european descent). Such method would improve the power of the statistical analyses as well as provide a better/equal representation of human genetics for genome-wide association studies.

#### Genetics and Genomics Research, 2021-2023
Statistical analysis on genetics and genomics data (transcriptomics and metabolomics so far) in context of Rheumatoid Arthritis treatment. Built a HLA analysis package written in python, [hapy](https://github.com/chuanfuyap/hla-analysis-py) as part of my research on HLA genetics analysis.
#### Metagenomics Research, 2019-2021
Studied metagenomics of lung microbiome in context of asthma using mathematical, statistical and machine learning models. Various scripts for data processing and analysis was turned into a python package, [motupy](https://github.com/chuanfuyap/motu-python-package)
#### Kinetic Modelling of Cell Metabolism,  2014-2018  
This was my PhD project where I optimised ODE models of yeast cell metabolism using genetic algorithm; input data included flux and proteomics data. [Convenient Modeller](https://github.com/chuanfuyap/Convenient-Modeller) was developed as part of this work.
## Education
#### PhD, University of Manchester,     2014-2018  
PhD in Bioinformatics, supervised by [Dr Jean-Marc Schwartz](https://www.research.manchester.ac.uk/portal/jean-marc.schwartz.html) and [Prof Simon Hubbard](https://www.research.manchester.ac.uk/portal/simon.hubbard.html)
#### MBiolSci, University of Sheffield,     2010-2014
Integrated Masters in Biochemistry, with final year project focusing on dynamic model of macrophage infection by _Staph aureus_. 
## Publication
> My list of publication in chronological order

1) [Model Parameterization with Quantitative Proteomics: Case Study with Trehalose Metabolism in Saccharomyces cerevisiae](https://doi.org/10.3390/pr9010139)

2) [Respiratory eukaryotic virome expansion and bacteriophage deficiency characterize childhood asthma](https://doi.org/10.1038/s41598-023-34730-7)

3) [A Proteomics study of Rheumatoid Arthritis patients on Etanercept identifies putative biomarkers associated with clinical outcome measures](https://doi.org/10.1093/rheumatology/kead321)

4) [HLA-DRB1 and HLA-DQA1 associated with immunogenicity to adalimumab therapy in patients with rheumatoid arthritis](http://dx.doi.org/10.1136/ard-2023-223955)
